### ISSUE TITLE
Build cache SFX support

### ISSUE THREAD

**Issue by jonricaurte** [systemjs/builder#296]
Here is my gulp task:

gulp.task('js', function() {

  console.log(new Date());
  var builder = new Builder({
    baseURL: '.',
    transpiler: 'babel'
  });

  builder.loadConfig('config.js')
    .then(function() {
      builder.buildSFX('app/assets/javascripts/application.js', 'public/assets/application.js')
        .then(function () {
          console.log(new Date());
          console.log('Build complete');
        })
        .catch(error);
    })
    .catch(error);

  function error(msg) {
    console.log('Build error');
    console.log(msg);
  }
});

It takes around 30 seconds for it to compile everything the first time, and around 20 seconds the second time. Is there any way to optimize this? I am using jspm and importing my third party references.

Thanks.

**Comment by guybedford** [comment]
How large is your build file here? If you want a faster build you can use the SystemJS builder cache API which is:

```javascript
var cache = builder.getCache();
builder.setCache(cache);
```

**Comment by jonricaurte** [comment]
Still takes around 20 seconds:

gulp.task('js', function() {
  console.log(new Date())
  var builder = new Builder({
    baseURL: '.',
    transpiler: 'babel'
  });

  var cache = builder.getCache();
  builder.setCache(cache);

  builder.loadConfig('config.js')
    .then(function() {
      builder.buildSFX('app/assets/javascripts/application.js', 'public/assets/application.js')
        .then(function () {
          console.log(new Date())
          console.log('Build complete');
        })
        .catch(error);
    })
    .catch(error);

  function error(msg) {
    console.log('Build error');
    console.log(msg);
  }
});

**Comment by guybedford** [comment]
How large is the output file? Are you using minification or not? There is a `lowResSourceMaps: true` build option which can speed up the source maps generation. Did you actually implement a cache here? Note the cache only benefits repeated builds.

**Comment by jonricaurte** [comment]
Where would you suggest putting the calls for caching? I put them in but it still does not seem to work.

**Comment by jonricaurte** [comment]
The output file is 2.2MB and I am not using minification. I also don't have sourcemaps but are they auto injected in the sfx build?

**Comment by guybedford** [comment]
For the cache you'd want to retrieve it after the old build via `getCache`, and then store it in memory until a future build, setting it with `setCache` on the new build before starting it.

**Comment by jonricaurte** [comment]
so here my updated one:

```
var cache;

gulp.task('js', function() {
  console.log(new Date())
  var builder = new Builder({
    baseURL: '.',
    transpiler: 'babel'
  });
  
  if (cache) {
    builder.setCache(cache);
  }
  
  builder.loadConfig('config.js')
    .then(function() {
      builder.buildSFX('app/assets/javascripts/application.js', 'public/assets/application.js')
        .then(function () {
          cache = builder.getCache();
          console.log(new Date())
          console.log('Build complete');
        })
        .catch(error);
    })
    .catch(error);

  function error(msg) {
    console.log('Build error');
    console.log(msg);
  }
});
```

and i get an error:

{ [Error: ENOENT, open '/Users/path/1.js']
  errno: 34,
  code: 'ENOENT',
  path: '/Users/path/1.js' }

**Comment by guybedford** [comment]
Thanks! This would be a buildSFX cache bug... I'll look into it as soon as I can.

**Comment by guybedford** [comment]
(up until now cached builds have only been tested against full bundles)

**Comment by jonricaurte** [comment]
Thanks!

**Comment by jonricaurte** [comment]
@guybedford was just wondering how soon do you think you will be able to have this fixed? I want to switch us over to a new build system using systemjs builder but with how long it is taking currently, I will have to push switching over out. Is there any time frame for when you might be able to fix this?

Thanks!

**Comment by guybedford** [comment]
This feature is now working on master. Aiming to get this major release out soon.

**Comment by jonricaurte** [comment]
Thanks for the update!

**Comment by guybedford** [comment]
Released in 0.14.0.
